### PR TITLE
Math unit test + Vec2 improvements

### PR DESCRIPTION
**Pull request by einarf**
* An attempt to make complete unit tests for Vec2
* Fix unsupported cases for operators in Vec2

I have not ran any performance tests in this yet. I reverse the if test in most of the operators attempting to subscript first and fall back to scalar. That also fixed a bug in `__mul__` were a vector would end up like this: `Vec2(x=Vec2, y=Vec2)`. Possibly it's faster to do `if hasattr(value, "__len__")` or some other crazy stuff instead of attempting the subscript.

It should not be too hard to add tests for the other vectors doing similar tests.

Possibly it would be more intuitive to use `length()` for the vector length and use `abs` for absolute values

```py
>> abs(Vec2(-1, -1)
Vec2(1, 1)
>> Vec2(1. 0).length()
1.0
```
Both of these are common use cases.

I can do more changes in this PR or start a new one.

### CODE REVIEW

**Comment by DragonMoffon** [comment]
I was under the impression that dot notation was faster than indexing into named tuples, but I ran some perf tests
```python
from pyglet.math import Vec2
from timeit import timeit
from random import random

def make_random():
    return Vec2(random(), random()), Vec2(random(), random())


def sum_a(vec1: Vec2, vec2: Vec2):
    Vec2(vec1[0] + vec2[0], vec1[1] + vec2[1])


def sum_b(vec1: Vec2, vec2: Vec2):
    Vec2(vec1.x + vec2.x, vec1.y + vec2.y)


n = 10000000
print(timeit(lambda: sum_a(*make_random()), number=n) / n * 1000)
print(timeit(lambda: sum_b(*make_random()), number=n) / n * 1000)
```

results
```
0.0007163603900000454 ms
0.0007484691700083204 ms
```
which implies that even for `NamedTuples` indexing is faster
So, the current code is the best option, especially for the flexibility of using standard tuples.

**Comment by einarf** [comment]
I read that as the new code being faster. The effect if reversing the `if` tests still needs to be looked at. It needs to happen anyway because `__mul__` would treat a `Vec2` as as a scalar because the exception doesn't trigger and you end up with `Vec2(x=Vec2(x, y), y=Vec2(x,y)` (something that is very cursed)

* Noticed I forgot to test swizzle.
* Will `index()` be called internally by something. Just asking to be sure because `NotImplemented` vs `NotImplementedError` has special meaning.

**Comment by pushfoo** [comment]
```
$ python bench_dragon_tuples.py
0.0009780469507999442
0.0010380777733000287
```
```console
$ python --version
Python 3.9.2
```

**Comment by benmoran56** [comment]
These benchmarks are consistently frustrating. 
I tested this extensively, and dot lookup on NamedTuple subclasses was consistently faster than indexed lookup. But now I test again today, and i'm getting similar results as here. The numbers don't lie though.
```sh
ben@ncase ~/Documents/PycharmProjects/pyglet $ python vectorbench-dragon.py 

function                            DotAddVec2   IndexAddVec2 change %    
---------------------------------------------------------------------------
Instantiation                       0.2109       0.21         0.44 %
Addition                            0.3663       0.3113       17.67 %
ben@ncase ~/Documents/PycharmProjects/pyglet $ python vectorbench-dragon.py 

function                            DotAddVec2   IndexAddVec2 change %    
---------------------------------------------------------------------------
Instantiation                       0.2069       0.207        -0.08 %
Addition                            0.3729       0.311        19.93 %
ben@ncase ~/Documents/PycharmProjects/pyglet $ python vectorbench-dragon.py 

function                            DotAddVec2   IndexAddVec2 change %    
---------------------------------------------------------------------------
Instantiation                       0.2088       0.2195       -4.86 %
Addition                            0.3631       0.3191       13.78 %
ben@ncase ~/Documents/PycharmProjects/pyglet $ 
```

**Review by benmoran56 — APPROVED** [review]

File: pyglet/math.py [hunk 1]
```diff
@@ -53,54 +53,85 @@ class Vec2(_typing.NamedTuple):
...
-            raise err
+            return Vec2(scalar[0] // self[0], scalar[1] // self[1])
+        except TypeError:
+            return Vec2(scalar // self[0], scalar // self[1])
 
-    __rsub__ = __sub__
-    __rmul__ = __mul__
-    __rtruediv__ = __truediv__
-    __rfloordiv__ = __floordiv__
+    __rmul__ = __mul__  # Order doesn't matter here so we can use __mul__
```
[the remark this answers is not in the document]
Comment (einarf) [reply pyglet/math.py:112]: The greedy approach only works for commutative operators meaning add and multiply. Without being explicit with sub and div you get the wrong ordering. See unit test for all the new unlocked possibilities with operators.

**Pull request merged**